Add bioclim

sdtaylor opened the pull request:
This script only requires the file .data/bbs_data.csv, which is loaded inside forecasting-bbs-core.R , and the appropriate packages. 

If you source this script and run the command `bbs_bioclim=get_bioclim_data()`, it will do all the legwork of downloading prism data, extracting values, and calculating bioclim variables. It will also store bioclim and prism data in .data/bbsforecasting.sqlite to make things quicker in future calls.

ethanwhite commented on get_prism_data.R:
```diff
@@ -4,39 +4,232 @@ library(prism)
+  database <- src_sqlite(sqliteDBFile, create = FALSE)
+} else {
+  database <- src_sqlite(sqliteDBFile, create = TRUE)
+}
```
I think we can just use `database <- src_sqlite(sqliteDBFile, create = TRUE)` with none of the conditional logic. The documentation is unclear on this point, but a quick test suggests that `create` only comes into play if the file doesn't exist already (and this would be a pretty common approach to this sort of API).

sdtaylor replied:
I would rather keep the conditional in case `create` changes in the future and, for example, throws an error if `create = TRUE` and the file already exists.

ethanwhite replied:
OK.

ethanwhite replied:
I've also opened a PR to see if they'll clarify this intended behavior for future use:
https://github.com/hadley/dplyr/pull/1664

ethanwhite replied:
FYI, this PR has been merged now, so we can count on `src_sqlite` to still read existing DBs even when `create = TRUE` from now on.

ethanwhite commented on get_prism_data.R:
```diff
@@ -4,39 +4,232 @@ library(prism)
+  } else {
+    return(FALSE)
+  }
+}
```
It seems like it would be cleaner to use `identical` to directly compare `prismLS` and `toCheck`. This would save both the merge and the summing and therefore the need to explain how all of this is working, and so could replace lines 66-75 with `return(identical(prismLS, toCheck))`

sdtaylor replied:
identical! awesome, didn't know that existed.

ethanwhite commented:
At the start of `get_prism_data()` `try` is used to check for the existence of the `prism_bbs_data`, i.e., to figure out if the database has been loaded or created. I think the logic could be made a little more straightforward here by using `prism_bbs_data %in% src_tbls(database)`.

ethanwhite commented on get_prism_data.R:
```diff
@@ -4,39 +4,232 @@ library(prism)
+#################################################################
+#From raw prism monthly values calculate all the bioclim variables.
+#You should not call this directly to load bioclim vars. Instead call
+#get_bioclim_data(), which will 1st try to load the data from the sdqlite
```
"sdqlite" -> "sqlite"

ethanwhite commented on get_prism_data.R:
```diff
@@ -4,39 +4,232 @@ library(prism)
+#From raw prism monthly values calculate all the bioclim variables.
+#You should not call this directly to load bioclim vars. Instead call
+#get_bioclim_data(), which will 1st try to load the data from the sdqlite
+#db before processing it all from scratch. 
```
This function documentation should be directly attached to the function and the helper function gets it own separate docs (which it basically already has).

ethanwhite commented on get_prism_data.R:
```diff
@@ -4,39 +4,232 @@ library(prism)
-  system(paste("raster2pgsql -s 4326", bil_file_path, datadir, ">", sql_file_path))
-  system(paste("psql -d bbsforecasting -f", sql_file_path))
+process_bioclim_data=function(){
+  #Get the prism data. It's pulled from the sqlite DB or downloaded as needed.
```
Comments like this are generally considered unnecessary because they just restate the information that you get by reading the code. If folks need to know where the PRISM data is coming from they can read `get_prism_data()`.

ethanwhite commented:
We generally want to try to avoid mixed naming conventions (e.g., pot_hole and camelCase for variables) in the same project. As a general style guide we tend to use [Hadley Wickham](http://r-pkgs.had.co.nz/style.html).

ethanwhite commented on get_prism_data.R:
```diff
@@ -4,39 +4,232 @@ library(prism)
+#load the prism data, process bioclim, load it in sqlite, then return 
+#bioclim as requested.
+
+#Possibly put a check here to make sure all years are accounted for?
```
I generally recommend starting comments like this with `TODO: `. This lets us quickly grep the code base for things that are marked in this way and lets us make sure that we've taken care of all of them before submitting. Alternatively they can be added as issues in the github repo. Likewise, if something is a fix, rather than a to do, it's common to use `FIXME: ` in an equivalent way.

ethanwhite commented on get_prism_data.R:
```diff
@@ -4,39 +4,232 @@ library(prism)
+
+get_bioclim_data=function(){
+  #Query the bioclim data and check to see if it throws an error. 
+  bioclim_bbs_data=try(collect(tbl(database, sql('SELECT * from bioclim_bbs_data'))))
```
Similar comment to before about explicitly checking to see if the table exists rather than catching an error and interpreting the type.

ethanwhite commented on get_prism_data.R:
```diff
@@ -4,39 +4,232 @@ library(prism)
+    print("bioclim data table not found, processing from scratch. ")
+    bioclim_bbs_data=process_bioclim_data()
+    
+    x=copy_to(database, bioclim_bbs_data, temporary = FALSE, 
```
Does this require assignment? Seems like you can probably just drop the `x=`.

ethanwhite commented:
This looks awesome @sdtaylor! Nice work.

Because it's in such great shape already I've made a bunch of comments related to style and making things easier to read. This is partly to help improve this particular code, but also just to start having a more general conversation about style and readability since your code is already so excellent.

Take what's useful and ping me when you're happy with your updates and we'll get this merged.

sdtaylor commented:
I implemented all your other suggests. 
-simplified checking if tables already exists by using `table_name %in% src_tbls(database)`
-organized the comments better and fixed some typos
-changed all camelCase to pot_hole to match the rest of the project. 

Still working on using identical() in checking for all the prism rasters being present. It's very picky about characters vs factors. 
all.equal() does the same thing but is less strict, so I may try that.

ethanwhite commented:
That all looks great.

> It's very picky about characters vs factors. 

Can you just set `stringsAsFactors = FALSE` in `expand.grid`.

sdtaylor commented:
setting `stringsAsFactors = FALSE` does not work either with `identical`. But it does make `as.equal()` do what I'm trying to do.

sdtaylor commented:
Last suggestion implemented. 
-Simplified check for all the prism rasters by using all.equal()

ethanwhite commented:
Looks great!
